### PR TITLE
add gcc attribute for json_sprintf and json_vsprintf

### PR DESCRIPTION
**Pull request by Mephistophiles**
Added gcc attribute for strict validation format string.

### CODE REVIEW

**Comment by coreyfarrell** [comment]
This is a good idea but I believe this attribute is supported by more than just gcc (for example clang/llvm).  I'm not sure if we should use the attribute for `#if defined(__GNUC__) || defined(__clang__)` or if we should add a check to configure.ac and CMakeLists.txt, then add `#define attribute_printf(fmt, args) @ATTRIBUTE_PRINTF@` to src/jansson_config.h.in.

@akheron what do you think?

**Comment by akheron** [comment]
In my opinion, the check for `__GNUC__` or `__clang__` is the way to go, no need to involve configure.ac or CMakeLists.txt.

File: src/jansson.h [hunk 1]
```diff
@@ -291,7 +291,14 @@ int json_vunpack_ex(json_t *root, json_error_t *error, size_t flags, const char
 
 /* sprintf */
 
+#ifdef __GNUC__
+__attribute__((format(printf, 1, 2)))
+#endif
```
Comment (akheron) [inline src/jansson.h]: Is this attribute correct before the function declaration? In the examples in https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html, attributes are always specified after the function declataion.

**Comment by coreyfarrell** [comment]
> In my opinion, the check for __GNUC__ or __clang__ is the way to go, no need to involve configure.ac or CMakeLists.txt.

Any opinion on creating a block that defines `JSON_ATTR_PRINTF(fmt, first_arg_idx)` to produce the attribute or nothing depending on the compiler?  This may seem like overkill for the 2 uses of this attribute type but other attributes would be useful to enhance the compiler warnings, for example `warn_unused_result` on `strbuffer_init` would get the compiler to warn us when ignore that allocation failure.  I think having the check for `__GNUC__` or `__clang__` on every function that uses an attribute would make the header more difficult to read.

**Comment by akheron** [comment]
> Any opinion on creating a block that defines JSON_ATTR_PRINTF(fmt, first_arg_idx) to produce the attribute or nothing depending on the compiler?

Go for it. Let's see how it looks :)

File: src/jansson.h (hunk 1 above)
Comment (Mephistophiles) [reply src/jansson.h]: Yes. You can use the attributes before and after declaring the function. If you want - I can move the attribute after the function declaration.

**Comment by Mephistophiles** [comment]
It was offered a more elegant solution, so I close this pull req.

**Pull request closed without merging**